tests/src/JIT/Performance/CodeQuality/Span/SpanBench.cs

```diff
@@ -447,82 +546,189 @@ public static void TestSpanDangerousGetPinnableReferenceByte(int length)
         [InlineData(10)]
         [InlineData(100)]
         [InlineData(1000)]
-        public static void TestSpanIndexString(int length)
+        public static void TestArrayIndexHoistableString(int length)
         {
-            var array = new string[length];
-            var span = new Span<string>(array);
-            string temp;
-            foreach (var iteration in Benchmark.Iterations)
-                using (iteration.StartMeasurement())
-                    for (int i = 0; i < Benchmark.InnerIterationCount; i++)
-                        temp = span[length / 2];
+            InvokeTestArrayIndexHoistable<string>(length);
+        }
+
+        static void InvokeTestArrayIndexHoistable<T>(int length)
+        {
+            var array = new T[length];
+            Invoke((int innerIterationCount) => TestArrayIndexHoistable<T>(array, length, innerIterationCount),
+                "TestArrayIndexHoistable<{0}>({1})", typeof(T).Name, length);
+        }
+
+        [MethodImpl(MethodImplOptions.NoInlining)]
+        static void TestArrayIndexHoistable<T>(T[] array, int length, int iterationCount)
+        {
+            var sink = Sink<T>.Instance;
+
+            for (int i = 0; i < iterationCount; i++)
+                sink.Data = array[length / 2];
         }
         #endregion
 
-        #region TestArrayIndex<T>
+        #region TestSpanIndexVariant<T>
         [Benchmark(InnerIterationCount = BaseIterations)]
         [InlineData(1)]
         [InlineData(10)]
         [InlineData(100)]
         [InlineData(1000)]
-        public static void TestArrayIndexByte(int length)
+        public static void TestSpanIndexVariantByte(int length)
         {
-            var array = new byte[length];
-            byte temp;
-            foreach (var iteration in Benchmark.Iterations)
-                using (iteration.StartMeasurement())
-                    for (int i = 0; i < Benchmark.InnerIterationCount; i++)
-                        temp = array[length / 2];
+            InvokeTestSpanIndexVariant<byte>(length);
+        }
 
+        [Benchmark(InnerIterationCount = BaseIterations)]
+        [InlineData(1)]
+        [InlineData(10)]
+        [InlineData(100)]
+        [InlineData(1000)]
+        public static void TestSpanIndexVariantString(int length)
+        {
+            InvokeTestSpanIndexVariant<string>(length);
+        }
+
+        static void InvokeTestSpanIndexVariant<T>(int length)
+        {
+            var array = new T[length];
+
+            Invoke((int innerIterationCount) => TestSpanIndexVariant<T>(array, length, innerIterationCount),
+                "TestSpanIndexVariant<{0}>({1})", typeof(T).Name, length);
         }
 
+        [MethodImpl(MethodImplOptions.NoInlining)]
+        static void TestSpanIndexVariant<T>(T[] array, int length, int iterationCount)
+        {
+            var sink = Sink<T>.Instance;
+            var span = new Span<T>(array);
+            int mask = (length < 2 ? 0 : (length < 8 ? 1 : 7));
```

Comment:
It follows the sequence 0,1,2,3,4,5,6,7,0,1,2,3,4,5,6,7,0,1,2,3,...  By "variant" here I just meant "not the same on each iteration of the loop" so that the JIT can't hoist the span access out of the loop.